libbeat/instrumentation/instrumentation.go

```diff
@@ -0,0 +1,244 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package instrumentation
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"go.elastic.co/apm"
+	apmtransport "go.elastic.co/apm/transport"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/transport"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+func init() {
+	// we need to close the default tracer to prevent the beat sending events to localhost:8200
+	apm.DefaultTracer.Close()
+}
+
+// Instrumentation is an interface that can return an APM tracer a net.listener
+type Instrumentation interface {
+	Tracer() *apm.Tracer
+	Listener() net.Listener
```

Comment:
I'm curious, but why do we need the Listener here? Is it still for APM itself, or is it some general requirement? I didn't see the listener used.

Reply:
Yes, it is for apm-server: we need to create an adhoc tracing server so it doesn't send traces to itself infinitely. Possibly this can be useful for the Elastic Agent too.